File: 0002-add-two-numbers/0002-add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummyHead = ListNode(0)
        curr = dummyHead
        carry = 0
        while l1 != None or l2 != None or carry != 0:
            l1val = l1.val if l1 else 0
            l2val = l2.val if l2 else 0
            col_val = l1val + l2val + carry
            carry = col_val // 10
            real_val = ListNode(col_val % 10)
            curr.next = real_val
            curr = curr.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return dummyHead.next

File: 0002-add-two-numbers/0002-add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        if(l1 == null || l2 == null) return null;
        ListNode res = new ListNode(-1);
        ListNode dummy = res;
        int carry = 0;
        while(l1 != null && l2 != null) {
            int sum = l1.val + l2.val + carry;
            int finalSum = sum % 10;
            carry = sum / 10;
            dummy.next = new ListNode(finalSum);
            dummy = dummy.next;
            l1 = l1.next;
            l2 = l2.next;
        }
        while(l1 != null) {
            int sum = l1.val + carry;
            int finalSum = sum % 10;
            carry = sum / 10;
            dummy.next = new ListNode(finalSum);
            dummy = dummy.next;
            l1 = l1.next;
        }
        while(l2 != null) {
            int sum = l2.val + carry;
            int finalSum = sum % 10;
            carry = sum / 10;
            dummy.next = new ListNode(finalSum);
            dummy = dummy.next;
            l2 = l2.next;
        }
        if(carry != 0) dummy.next = new ListNode(carry);
        return res.next;
    }
}